src/traces/sunburst/calc.js

```diff
@@ -0,0 +1,237 @@
+        else parent2children[parent] = [v];
+        refs[v] = 1;
+    };
+
+    var isValidVal = function(i) {
+        return !hasVals || (isNumeric(vals[i]) && vals[i] >= 0);
+    };
+
+    var len;
+    var isValid;
+    var getId;
+
+    if(hasIds) {
+        len = Math.min(ids.length, parents.length);
+        isValid = function(i) { return ids[i] && isValidVal(i); };
+        getId = function(i) { return String(ids[i]); };
+    } else {
+        len = Math.min(labels.length, parents.length);
+        isValid = function(i) { return labels[i] && isValidVal(i); };
+        // TODO We could allow some label / parent duplication
+        //
+        // From AJ:
+        //  It would work OK for one level
+        //  (multiple rows with the same name and different parents -
+        //  or even the same parent) but if that name is then used as a parent
+        //  which one is it?
+        getId = function(i) { return String(labels[i]); };
+    }
+
+    if(hasVals) len = Math.min(len, vals.length);
+
+    for(var i = 0; i < len; i++) {
+        if(isValid(i)) {
+            var id = getId(i);
+            var pid = parents[i] ? String(parents[i]) : '';
+
+            var cdi = {
+                i: i,
+                id: id,
+                pid: pid,
+                label: labels[i] ? String(labels[i]) : ''
+            };
+
+            if(hasVals) cdi.v = +vals[i];
+            cd.push(cdi);
+            addToLookup(pid, id);
+        }
+    }
+
+    if(!parent2children['']) {
+        var impliedRoots = [];
+        var k;
+        for(k in parent2children) {
+            if(!refs[k]) {
+                impliedRoots.push(k);
+            }
+        }
+
+        // if an `id` has no ref in the `parents` array,
+        // take it as being the root node
+
+        if(impliedRoots.length === 1) {
+            k = impliedRoots[0];
+            cd.unshift({
+                id: k,
+                pid: '',
+                label: k
+            });
+        } else {
+            return Lib.warn('Multiple implied roots, cannot build sunburst hierarchy.');
+        }
+    } else if(parent2children[''].length > 1) {
+        var dummyId = Lib.randstr();
+
+        // if multiple rows linked to the root node,
+        // add dummy "root of roots" node to make d3 build the hierarchy successfully
+
+        for(var j = 0; j < cd.length; j++) {
+            if(cd[j].pid === '') {
+                cd[j].pid = dummyId;
+            }
+        }
+
+        cd.unshift({
+            hasMultipleRoots: true,
+            id: dummyId,
+            pid: ''
+        });
+    }
+
+    // TODO might be better to replace stratify() with our own algorithm
+    var root;
+    try {
+        root = d3Hierarchy.stratify()
+            .id(function(d) { return d.id; })
+            .parentId(function(d) { return d.pid; })(cd);
+    } catch(e) {
+        return Lib.warn('Failed to build sunburst hierarchy.');
```

Comment:
It might be useful to return the error message from `stratify()`:
```
return Lib.warn('Failed to build sunburst hierarchy with error: `' + e.message + '`');
```
to help understand why a given dataset is failing to render.

Reply:
Good call! Done in https://github.com/plotly/plotly.js/pull/3594/commits/4a2f7a5a59e7868b1a21222647d2632cb7d3edb0